crypto/s2n_certificate.c

```diff
@@ -238,8 +268,57 @@ int s2n_send_cert_chain(struct s2n_stuffer *out, struct s2n_cert_chain *chain)
     return 0;
 }
 
-int s2n_send_empty_cert_chain(struct s2n_stuffer *out) {
+int s2n_send_empty_cert_chain(struct s2n_stuffer *out)
+{
     notnull_check(out);
     GUARD(s2n_stuffer_write_uint24(out, 0));
     return 0;
 }
+
+static int s2n_does_cert_san_match_hostname(X509 *cert, const char *hostname)
+{
+    GENERAL_NAMES *san_names = NULL;
+    if (!(san_names = X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL))) {
+        /* No SAN extension */
+        return 0;
+    }
+
+    const int hostname_len = strnlen(hostname, S2N_MAX_SERVER_NAME);
+    const int num_san_names = sk_GENERAL_NAME_num(san_names);
+    for (int i = 0; i < num_san_names; i++) {
+        GENERAL_NAME *san_name = sk_GENERAL_NAME_value(san_names, i);
+        if (!san_name) {
+            continue;
+        }
+
+        /* we only care about DNS entries */
+        if (san_name->type == GEN_DNS) {
+            unsigned char *san_str = NULL;
+            const int san_str_len = ASN1_STRING_to_UTF8(&san_str, san_name->d.dNSName);
+            if (san_str_len < 0) {
+                /* Some error parsing the SAN, skip the match and don't free memory. */
+                continue;
+            }
+
+            const int match = !!((hostname_len == san_str_len) && (strncasecmp(hostname, (const char *) san_str, san_str_len) == 0));
```

Comment:
Do we need to worry about wildcard certs here? Eg Certificate has `*.animals.com`, but SNI is `alligator.animals.com`?

Reply:
Definitely -- I'm currently planning on knocking that out in a separate PR. I'm thinking this PR will just cover the most basic name lookup functionality.